GH-3869: Prune empty queues

mpapierski opened the pull request:
Closes #3869

This PR adds an ability to purge `Key::Withdraw` entries after upgrading to the `Key::Unbond` keyspace and also purges empty queues in the `Key::Unbond` space to reduce the width of a tip of a Merkle trie.

This is achieved by introducing a `Transform::Delete` operation which is emitted on each auction's step, which is then committed and an entry is purged from the new tip of the trie. New tracking copy operation "delete" is private, and is not exposed to the public, as it's only used within the system contract.

Commit:
Fix clippy issue.

`rlimit` module handles linux/macos differencies, so removing this line
makes clippy happy.

Commit:
Add "Delete" transform.

This adds a "Delete" transform, which when applied, removes a key from
the tip of a trie.

Commit:
Remove unused Withdraw/Unbond entries.

After performing initial upgrade from Withdraw -> Unbond key space,
withdraw keys are purged from the tip of the trie. After that, once an
unbonding queue is empty, then given entry is also removed from the
trie.

Commit:
Clean ups and more tests.

Fraser999 commented on execution_engine/src/shared/transform.rs:
```diff
@@ -168,23 +170,42 @@ impl Transform {
+            }
+            Err(error) => Err(error),
+        }
+    }
+    /// Applies the transformation on a specified stored value instance.
```
```suggestion
    }

    /// Applies the transformation on a specified stored value instance.
```

Fraser999 commented on execution_engine/src/shared/transform.rs:
```diff
@@ -168,23 +170,42 @@ impl Transform {
+    /// This method produces a new [`StoredValue`] instance based on the [`Transform`] variant. If a
+    /// given transform is a [`Transform::Delete`] then `None` is returned as the [`StoredValue`] is
+    /// consumed but no new value is produced.
+    fn apply_optional(self, stored_value: StoredValue) -> Result<Option<StoredValue>, Error> {
```
I don't think we should have a separate method only to handle `Transform::Delete`s without panicking.

I think it would be much clearer if `apply` were just updated to return the optional stored value and we updated the few callsites to handle this.  Probably in many(all?) cases we could just panic at the callsite if `Ok(None)` is returned, but I feel this is more future-proof/less footgunny than putting the panicking code inside `apply`.

mpapierski replied:
Done as discussed: `Transform::apply` now returns `Option<StoredValue>` where Some indicate newly produced value, and None means no new value was produced and the value should be deleted from the global state.

Fraser999 commented on execution_engine_testing/tests/src/test/system_contracts/auction/bids.rs:
```diff
@@ -3548,6 +3552,35 @@ fn should_continue_auction_state_from_release_1_4_x() {
+        .iter()
+        .all(|key| key.as_withdraw().is_some()));
+
+    // Ensure withdraw keys are purged
```
```suggestion
    // Ensure withdraw keys are pruned
```

mpapierski replied:
Fixed

Fraser999 commented on execution_engine_testing/tests/src/test/system_contracts/auction/bids.rs:
```diff
@@ -3717,6 +3750,22 @@ fn should_continue_auction_state_from_release_1_4_x() {
+    assert_eq!(
+        unbond_keys.len(),
+        0,
+        "auction state continued and empty unbond queue should be purged"
```
```suggestion
        "auction state continued and empty unbond queue should be pruned"
```

mpapierski replied:
Fixed

Fraser999 commented on node/src/reactor.rs:
```diff
@@ -58,7 +58,6 @@ use tokio::time::{Duration, Instant};
 use tracing::{debug, debug_span, error, info, instrument, trace, warn, Span};
 use tracing_futures::Instrument;
 
-#[cfg(target_os = "linux")]
```
Removed by mistake?

mpapierski replied:
Intentional, actually, as currently build fails on macOS

Fraser999 replied:
Hmm, seems like `#[cfg(target_os = "linux")]` has been removed from `TARGET_OPEN_FILES_LIMIT` and `fn adjust_open_files_limit()` (compare to [this](https://github.com/casper-network/casper-node/blob/039d438f2c7b15a007e9248154d6f955bf63b740/node/src/reactor.rs#L109) and [this](https://github.com/casper-network/casper-node/blob/039d438f2c7b15a007e9248154d6f955bf63b740/node/src/reactor.rs#L113)).

Could you change these back to include the gating please?  (It was probably me who broke this!)

mpapierski replied:
Tried my best to revert as closely as possible here 6af61de22e0b31f0dd186215e2a07cf30aa55596

Fraser999 commented on types/src/execution_result.rs:
We should update `impl Distribution<Transform> for Standard` too to include `Transform::Delete`

mpapierski replied:
Done

Fraser999 reviewed (commented):
Mostly nitpicks, but generally looks good to me.  Main concern is the panicking code added to `Transform::apply`.

Commit:
Apply @Fraser999's comments

This changes transform API to return optional that indicates new value
(if Some) or a deletion (if None).

Commit:
Update schemas and fix tests

Commit:
Revert conditional compilation for rlimits.

real-felix commented on execution_engine_testing/tests/src/test/system_contracts/auction_bidding.rs:
```diff
@@ -540,10 +538,189 @@ fn should_run_successful_bond_and_unbond_with_release() {
+    );
+}
+
+#[ignore]
```
Maybe there should be an explanation about why it's ignored?

mpapierski replied:
This is our convention for integration tests, so `cargo test` runs unit tests (fast), and running ignored tests runs the integration test suite (could be long)

Commit:
Merge branch 'feat-1.6' into gh-3869-prune-empty-queues

 Conflicts:
	node/src/utils.rs

rafal-ch commented on execution_engine/src/core/runtime/auction_internal.rs:
```diff
@@ -98,12 +98,15 @@ where
-            .map_err(|exec_error| <Option<Error>>::from(exec_error).unwrap_or(Error::Storage))
+        let unbond_key = Key::Unbond(account_hash);
+        if unbonding_purses.is_empty() {
+            self.context.delete_gs_unsafe(unbond_key);
```
Deleting a non-existing key is a noop, not a crash, right?

I noticed the `Use with caution` comment, which indicates that at some point someone will use it without caution :)

mpapierski replied:
Actually deleting nonexisting is not a noop, it's an error. We first need to read the StoredValue under a Key, then we'll apply a transform on it which could result a Some(new_stored_value) (write, add, identity, etc) or None (delete). Then persisting stored values would either write new data to global state (`Some`) or delete it (`None`)

> Unsafe

It's a convention we have, for storage APIs that are called directly from system contracts. It's not unsafe as in rust's unsafe keyword, it's just the fact that there's no validation, and we don't want to expose this API to users yet, so most if not all the methods I introduced are `pub(crate)` at most.

rafal-ch reviewed (approved):
LGTM :+1: 
with once question
